File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        //time: O(n * klog(k)) k is max str length
        //space: O(n * k) n unique keys of length k
        //map - key anagram, value original string
        Map<String, List<String>> map = new HashMap<>();
        for(String str : strs){
            char []c = str.toCharArray(); //sort letters to get anagram
            Arrays.sort(c);
            String key = new String(c); 
            if(map.containsKey(key)){
                map.get(key).add(str); //append to list
            }else{
                List<String> list= new ArrayList<>(); //create new list
                list.add(str);
                map.put(key,list);
            }
        }
        
        return new ArrayList<>(map.values()); //use parenthesis to convert array to arraylist
    }
}

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # map with key anagram and value array        
        res = {}
        
        for curr_str in strs:
            anagram = tuple(sorted(curr_str))  # Use tuple instead of list
            if anagram in res:
                res[anagram].append(curr_str)
            else:
                res[anagram] = [curr_str]

        return list(res.values())  # Return the list of grouped anagrams